src/software/backend/simulation/simulator_robot_singleton.h

```diff
@@ -0,0 +1,257 @@
+        app_wheel_destroy(back_left_wheel);
+
+        Wheel_t* back_right_wheel = app_firmware_robot_getBackRightWheel(firmware_robot);
+        app_wheel_destroy(back_right_wheel);
+
+        Wheel_t* front_right_wheel =
+            app_firmware_robot_getFrontRightWheel(firmware_robot);
+        app_wheel_destroy(front_right_wheel);
+
+        Chicker_t* chicker = app_firmware_robot_getChicker(firmware_robot);
+        app_chicker_destroy(chicker);
+
+        Dribbler_t* dribbler = app_firmware_robot_getDribbler(firmware_robot);
+        app_dribbler_destroy(dribbler);
+
+        ControllerState_t* controller_state =
+            app_firmware_robot_getControllerState(firmware_robot);
+        delete controller_state;
+
+        app_firmware_robot_destroy(firmware_robot);
+    };
+};
+
+
+/**
+ * This class acts as a wrapper around a SimulatorRobot so that the SimulatorRobot
+ * can provide the interface of a FirmwareRobot.
+ *
+ * Because our firmware structs rely on C-style function pointers, we
+ * cannot use C++ constructs like std::function to provide to the structs.
+ * Therefore we use this class's static functions to provide the function
+ * pointers, since static functions work as C-style function pointers
+ * (because there is no instance associated with them).
+ *
+ * Whenever a caller needs to get a firmware struct or perform an operation, they
+ * need to set which robot they want to control so that all the static functions
+ * that have been provided to the firmware struct operate on the correct
+ * instantiated object. This is our workaround to maintain and simulate multiple
+ * "instances" of robot firmware at once.
+ */
+class SimulatorRobotSingleton
+{
+   public:
+    /**
+     * Sets the SimulatorRobot being controlled by this class
+     *
+     * @param robot The SimulatorRobot being controlled by this class
+     */
+    static void setSimulatorRobot(std::shared_ptr<SimulatorRobot> robot);
+
+    /**
+     * Creates a FirmwareRobot_t with functions bound to the static functions in this
+     * class. Only one FirmwareRobot_t needs to be created to control all robots, since
+     * calling setSimulatorRobot will simply change the implementations of the bound
+     * functions to act as if the new robot was being controlled.
+     *
+     * @return a FirmwareRobot_t that is bound to whatever SimulatorRobot this Singleton
+     * is controlling
+     */
+    static std::unique_ptr<FirmwareRobot_t, FirmwareRobotDeleter> createFirmwareRobot();
+
+   private:
+    /** * Returns the x-position of the robot, in global field coordinates, in meters
+     *
+     * @return the x-position of the robot, in global field coordinates, in meters
+     */
+    static float getPositionX();
+
+    /**
+     * Returns the y-position of the robot, in global field coordinates, in meters
+     *
+     * @return the y-position of the robot, in global field coordinates, in meters
+     */
+    static float getPositionY();
```

Comment:
Are we able to just return the position as a point? It seems like every field is a primitive type, is this something to do with the firmware?

Reply:
We can not in this case because we need to provide these functions specifically when we create a `FirmwareRobot_t`. See how it's used in the `createFirmwareRobot` function.